src/client/api/event.c

```diff
@@ -1243,12 +1244,29 @@ daos_event_priv_wait()
 
 	/* Wait on the event to complete */
 	while (evx->evx_status != DAOS_EVS_READY) {
+		int rc2;
+
 		rc = crt_progress_cond(evx->evx_ctx, 0, ev_progress_cb, &epa);
-		if (rc == 0)
+
+		/** progress succeeded, loop can exit if event completed */
+		if (rc == 0) {
 			rc = ev_thpriv.ev_error;
+			continue;
+		}
 
-		if (rc && rc != -DER_TIMEDOUT)
-			break;
+		/** progress timeout, try calling progress again */
+		if (rc == -DER_TIMEDOUT)
+			continue;
+
+		/*
+		 * other progress failure; op should fail with that err. reset
+		 * the private event first so it can be resused.
+		 */
+		rc2 = daos_event_priv_reset();
```

Comment:
in dc_task_schedule, whatever we return gets ignored here

21         if (daos_event_is_priv(ev)) {
122                 daos_event_priv_wait();
123                 rc = ev->ev_error;
124         }

Do we need to make sure ev->ev_error is set?

Reply:
in success scenario, ev_error will be set.
in fail scenario, we reset the event, and task_schedule will return the crt_progress error reported, not the ev_error which is what we said we need to do in this scenario.

Reply:
it sets rc unconditionally to ev->ev_error after returning from daos_event_priv_wait().

Should it check the return value of daos_event_priv_wait() first ?

Reply:
nevermind, for some reason I was looking at local function rather than the one in your patch.